pumpkin-protocol/src/client/play/c_commands.rs

```diff
@@ -0,0 +1,382 @@
+use pumpkin_macros::client_packet;
+
+use crate::{bytebuf::ByteBuffer, ClientPacket, VarInt};
+
+#[client_packet("play:commands")]
+pub struct CCommands<'a> {
+    pub nodes: Vec<ProtoNode<'a>>,
+    pub root_node_index: VarInt,
+}
+
+impl<'a> CCommands<'a> {
+    pub fn new(nodes: Vec<ProtoNode<'a>>, root_node_index: impl Into<VarInt>) -> Self {
+        Self {
+            nodes,
+            root_node_index: root_node_index.into(),
+        }
+    }
+}
+
+impl<'a> ClientPacket for CCommands<'a> {
+    fn write(&self, bytebuf: &mut ByteBuffer) {
+        bytebuf.put_list(&self.nodes, |bytebuf, node: &ProtoNode| {
+            node.write_to(bytebuf)
+        });
+        bytebuf.put_var_int(&self.root_node_index);
+    }
+}
+
+pub struct ProtoNode<'a> {
+    pub children: Vec<VarInt>,
+    pub node_type: ProtoNodeType<'a>,
+}
+
+#[derive(Debug)]
+pub enum ProtoNodeType<'a> {
+    Root,
+    Literal {
+        name: &'a str,
+        is_executable: bool,
+    },
+    Argument {
+        name: &'a str,
+        is_executable: bool,
+        parser: ProtoCmdArgParser<'a>,
+        override_suggestion_type: Option<ProtoCmdArgSuggestionType>,
+    },
+}
+
+impl<'a> ProtoNode<'a> {
+    const FLAG_IS_EXECUTABLE: i8 = 4;
+    const FLAG_HAS_REDIRECT: i8 = 8;
+    const FLAG_HAS_SUGGESTION_TYPE: i8 = 16;
```

Comment:
Im not sure. but i think we just can make all these const's unsigned

Reply:
I don't think that's a good idea, because 1. I'm not sure either and 2. it does work this way and we don't really have a reason to change it and 3. it's consistent with vanilla which doesn't have unsigned integers

Reply:
1. Its should work.
2. When we are talking about unsigned, I don't want to keep consistency with Mojang. Because first there is no unsigned in Java :D. And Secound there are alot of cases which do not make sense to be signed and can actually break the game. For example the render distance is signed, Pumpkin just converts it into unsigned

Reply:
You're right actually. I for some reason completely ignored the fact that all that matters for flags is the binary representation